### PR TITLE
feat: Clawback Continuous Vesting Account

### LINKED ISSUE #194: ✨ Implement clawback continuous vesting account
## ✨ Feature

Implement a custom vesting account called `ClawbackContinuousVestingAccount` which wraps the SDK's `ContinuousVestingAccount` to provide an additional feature for cancelling a vesting stream.

- `tx create-vesting-account [to_address] [amount] [end_time]` for creating a `ClawbackContinuousVestingAccount`
  - The account must not already exist
  - The "from address" is registered as the newly created vesting account's funder, which is stored in the module keeper.

- `tx clawback [address]`
  - The funds that are still vesting are sent back to the funder's address
  - The funder corresponding to the vesting account is removed from the vesting module's store.

### PR DESCRIPTION
**Pull request by hacheigriega**
## Motivation

This PR implements a custom vesting account type `ClawbackContinuousVestingAccount` that wraps the SDK's `ContinuousVestingAccount` to provide an additional feature of "clawing back" funds to the funder of the account. Note that now this custom type is the only vesting account type our chain supports. 

## Explanation of Changes

The sender of the `create-vesting-account` transaction is registered as the `funder` of the vesting account so that when the `funder` later sends a `clawback` transaction, the vesting (= unvested) funds are returned back to the `funder`. The implementation is complicated by the fact that the funds can be in various states. The funds are clawed-back from different sources in the following order: Unbonded (through standard bank module transfer) -> Unbonding delegations -> Bonded delegations (which may also be going through redelegation process). The transfer logic is taken from Agoric's [`TransferDelegation`](https://github.com/agoric-labs/cosmos-sdk/blob/f42d86980ddfc07869846c391a03622cbd7e9188/x/staking/keeper/delegation.go#L701) and [`TransferUnbonding`](https://github.com/agoric-labs/cosmos-sdk/blob/f42d86980ddfc07869846c391a03622cbd7e9188/x/staking/keeper/delegation.go#L979). To integrate these staking keeper methods, we wire to the app the custom staking module that wraps the SDK's staking module. However, as evident in `x/staking/module.go`, the SDK's default staking handles the remaining functionalities of the staking module, and most of the code present in the custom staking module such as the `create-validator-vrf` transaction remains unactivated.

The clawed back amount is guaranteed to be all of the currently vesting amount, unless the vesting account's delegation has been slashed, and the account doesn't have sufficient unbonded funds to make up for the loss.

How to use the transactions:
- `tx create-vesting-account [to_address] [amount] [end_time]` for creating a `ClawbackContinuousVestingAccount`
  - The "to" account must not already exist.
  - The "from" address is registered as the newly created vesting account's funder.
  - Add a flag `--disable-clawback` to disallow clawback. If this flag is present, the `funder` field of the vesting account will be empty.

- `tx clawback [address]`
  - The funds that are still vesting are sent back to the funder's address.

⚠️Note the simulation testing is still being fixed - There is an issue with the simulation app state not being updated. 

## Testing

Integration testing and simulation testing

## Related PRs and Issues

Closes #194

### CODE REVIEW

**Review by gluax — COMMENTED** [review]
Just commenting since I'm not a cosmos expert.

It looks like we are replacing the original ContinousVestingAccoutns with Clawbackable ones. I thought we wanted both, more so for users than us; our personal main priority is Clawbacks for launch.

As for the legal stuff bc of licensing clash @jasperdg @mariocao I think GPLv3 is compatible with Apache 2. I believe this means we need to also include the Apache license. But might need to do more than that...

**Comment by hacheigriega** [comment]
> Just commenting since I'm not a cosmos expert.
> 
> It looks like we are replacing the original ContinousVestingAccoutns with Clawbackable ones. I thought we wanted both, more so for users than us; our personal main priority is Clawbacks for launch.
> 
> As for the legal stuff bc of licensing clash @jasperdg @mariocao I think GPLv3 is compatible with Apache 2. I believe this means we need to also include the Apache license. But might need to do more than that...

Good points. We can support both SDK's `ContinousVestingAccount` and our `ClawbackContinousVestingAccount`if we want. Not sure why we would want that though - just thought it would be simpler to support the clawback one only since they're identical except the clawback feature.

File: proto/sedachain/vesting/v1/tx.proto [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import "gogoproto/gogo.proto";
+import "cosmos/base/v1beta1/coin.proto";
+import "cosmos_proto/cosmos.proto";
+import "cosmos/vesting/v1beta1/vesting.proto";
+import "cosmos/msg/v1/msg.proto";
+import "amino/amino.proto";
+
+option go_package = "github.com/sedaprotocol/seda-chain/x/vesting/types";
+
+// Msg defines the bank Msg service.
```
[the remark this answers is not in the document]
Comment (hacheigriega) [reply proto/sedachain/vesting/v1/tx.proto:13]: ty

File: proto/sedachain/vesting/v1/tx.proto [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  // CreateVestingAccount creates a new vesting account.
+  rpc CreateVestingAccount(MsgCreateVestingAccount) returns (MsgCreateVestingAccountResponse);
+
+  // Clawback returns the vesting funds back to the funder.
+  rpc Clawback(MsgClawback) returns (MsgClawbackResponse);
+}
+
+// MsgCreateVestingAccount defines a message that creates a vesting account.
+message MsgCreateVestingAccount {
```
[the remark this answers is not in the document]
Comment (hacheigriega) [reply proto/sedachain/vesting/v1/tx.proto:25]: Currently it cannot create an original continuous vesting account. Adding that would be quite simple though, as we can just add a boolean field to the message to indicate clawback or not.

[the remark this answers is not in the document]
Comment (gluax) [reply proto/sedachain/vesting/v1/tx.proto:25]: Honestly, I'm not positive if we want it or not, but I figured from a user perspective, to be able to do both might be nice. idk thoughts @mariocao?

**Comment by jim380** [comment]
Currently `MsgClawback` fails in ABCI, particularly in `RunTx` when it’s called in `DeliverTx` in `execModeFinalize` mode.

If you run breakpoints or add a panic like below in `Clawback`, you will see that it never gets hits.
image

This confirms that the operation had failed in `RunTx` because msgs only get routed to the module protobuf msg services after a successful `RunTx`. Cc: @hacheigriega

File: proto/sedachain/vesting/v1/tx.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (mariocao) [reply proto/sedachain/vesting/v1/tx.proto:25]: If it is so simple, yes, why not :)

File: x/vesting/keeper/msg_server.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	if m.bk.BlockedAddr(to) {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", msg.ToAddress)
+	}
+
+	if acc := m.ak.GetAccount(ctx, to); acc != nil {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "account %s already exists", msg.ToAddress)
+	}
+
+	baseAccount := authtypes.NewBaseAccountWithAddress(to)
+	baseAccount = m.ak.NewAccount(ctx, baseAccount).(*authtypes.BaseAccount) // TO-DO what is this
```
Comment (mariocao) [inline x/vesting/keeper/msg_server.go:68]: TODO?

File: proto/sedachain/vesting/v1/tx.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (hacheigriega) [reply proto/sedachain/vesting/v1/tx.proto:25]: added a flag for disabling clawback.

**Review by gluax — APPROVED** [review]

File: x/vesting/keeper/msg_server.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+	}
+
+	bonded := sdk.NewCoins(sdk.NewCoin(bondDenom, bondedAmt))
+	unbonding := sdk.NewCoins(sdk.NewCoin(bondDenom, unbondingAmt))
+	unbonded := m.bk.GetAllBalances(ctx, vestingAccAddr)
+	total := bonded.Add(unbonding...).Add(unbonded...)
+
+	toClawBack := coinsMin(vestingAccount.GetVestingCoins(ctx.BlockTime()), total) // might have been slashed
+
+	// Write now now so that the bank module sees unvested tokens are unlocked.
```
Comment (jim380) [inline x/vesting/keeper/msg_server.go:166]: duplicate `now`

File: x/vesting/keeper/msg_server.go [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+// 1. vesting funds that have not been used towards delegation
+// 2. delegations
+// 3. unbonding delegations
+// in this order, as much as possible, until the vesting amount is met. Note that
+// due to slashing the funds from all three of these sources may still fail to meet
+// the vesting amount according to the vesting schedule.
+func (m msgServer) Clawback(goCtx context.Context, msg *types.MsgClawback) (*types.MsgClawbackResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// NOTE: errors checked during msg validation
```
Comment (jim380) [inline x/vesting/keeper/msg_server.go:113]: Where is this msg validation?

**Review by jim380 — COMMENTED** [review]
Looks great. Particularly the core logic in `func (m msgServer) Clawback`.

**Pull request merged**